### ISSUE TITLE
Unit tests are not passing

### ISSUE THREAD

**Issue by wkozaczuk** [cloudius-systems/osv#779]
Last week I noticed that tracing_smoke_test would fail every time I did run 'make check' which I thought maybe was a fluke or some local issue on my side. This however seems to be consistent with the results of latest OSv jenkins build (http://jenkins.cloudius-systems.com:8080/job/osv-build/1101/consoleFull and search for "Test tracing_smoke_test FAILED"). 

So actually the biggest problem is that Jenkins build is somehow misconfigured and does NOT report unit tests failing which is bad. I actually tried to find last jenkins run when tests were passing and I could not find one since May 23 2016 (oldest one in Jenkins history). 

Locally when I disabled tracing_smoke_test test I would get further to another test tst-regex.so consistenly failing: 
PASS: second match end-offset
terminate called after throwing an instance of 'std::regex_error'
  what():  regex_error
Aborted

[backtrace]
0x000000000048d744 <__gnu_cxx::__verbose_terminate_handler()+340>
0x0000000000726f71 <???+7499633>
Test tst-regex.so FAILED
Traceback (most recent call last):
  File "./scripts/test.py", line 170, in <module>
    main()
  File "./scripts/test.py", line 157, in main
    run_tests()
  File "./scripts/test.py", line 148, in run_tests
    run(tests_to_run)
  File "./scripts/test.py", line 68, in run
    run_test(test)
  File "./scripts/test.py", line 43, in run_test
    test.run()
  File "/home/ubuntu/projects/osv/scripts/tests/testing.py", line 29, in run
    run_command_in_guest(self.command).join()
  File "/home/ubuntu/projects/osv/scripts/tests/testing.py", line 157, in join
    raise Exception('Guest failed')
Exception: Guest failed
++ echo './scripts/build failed: ./scripts/test.py'
./scripts/build failed: ./scripts/test.py
++ exit 0

What was also disturbing was that occasionally I saw another test tst-namespace.so fail at least once in random fashion (meaning most of the time it would pass):

  TEST tst-namespace.so                   OSv v0.24-148-gfdcae29
eth0: 192.168.122.15
Assertion failed: sched::preemptable() (arch/x64/mmu.cc: page_fault: 33)

[backtrace]
0x0000000000225df8 <__assert_fail+24>
0x0000000000394abd <page_fault+269>
0x0000000000393906 <???+3750150>
0x00000000003f0e4d <sched::cpu::do_idle()+125>
0x00000000003f28a7 <sched::cpu::idle()+39>
0x00000000003f28ce <???+4139214>
0x00000000003f2b07 <thread_main_c+39>
0x0000000000394885 <???+3754117>
Test tst-namespace.so FAILED
Traceback (most recent call last):
  File "./scripts/test.py", line 170, in <module>
    main()
  File "./scripts/test.py", line 157, in main
    run_tests()
  File "./scripts/test.py", line 148, in run_tests
    run(tests_to_run)
  File "./scripts/test.py", line 68, in run
    run_test(test)
  File "./scripts/test.py", line 43, in run_test
    test.run()
  File "/home/ubuntu/projects/osv/scripts/tests/testing.py", line 29, in run
    run_command_in_guest(self.command).join()
  File "/home/ubuntu/projects/osv/scripts/tests/testing.py", line 157, in join
    raise Exception('Guest failed')
Exception: Guest failed
++ echo './scripts/build failed: ./scripts/test.py'
./scripts/build failed: ./scripts/test.py
++ exit 0

**Comment by wkozaczuk** [comment]
Honestly I am not sure if these tests are truly unit tests or more integration tests and how much isolated their runs are. In theory unit tests should not affect each other but given it is probably very hard to unit tests something OSv it maybe close to impossible. For example it is not clear to me if every single test runs in "fresh" OSv instance which would be full setUp and teardown or is single OSv instance reused across all test runs would mean that possibly memory state change by one unit test could affect another unit tests. But I do not know enough about OSv unit tests so I am just speculating here.

**Comment by nyh** [comment]
Hi, thanks for your analysis.

I think the "tracing_smoke_test" problem you are seeing might be the same issue as #772, which is not an OSv test failing, but rather some gdb commands which the test tries to use from an external script stopped working as libstdc++ changed. Issue #772 has a bit more detail on my guess as to what's causing this, but the failure in Jenkins looks a bit different (although again gdb related).

You're right that it's a second, separate bug that test.py reports an error and yet does "exit 0", and indeed this causes Jenkins not to report this failure. We'll need to fix that.

Regarding a tst-regex.so failure, I cannot reproduce this on my Fedora 23 with gcc 5.3.1 - it always succeeds here. What system and version of gcc are you building this test on?

You're right that  tst-namespace.so failing randomly is disturbing. I just ran it 200 times on my setup now, and it didn't fail in any of the attempts. So if you can fairly easily reproduce this failure, I would be grateful if you could attach to it with gdb (see https://github.com/cloudius-systems/osv/wiki/Debugging-OSv) and find a stack trace which is usually much more informative than that stack trace printed by OSv itself.

Regarding your last question: Currently, each of the test programs indeed runs on a fresh instance of OSv, in a fresh VM which boots, starts the OSv kernel, runs the single test shared object (e.g., tst-namespace.so) and shuts down the VM. This slows down the test process, but not by much (considering that OSv boots and shuts down in less than one second). You're right that had we tried to run all the tests one one VM, in parallel or in sequence, it would have been harder to isolate problems or debug where they come from (e.g., we could have the 10th test fail just because the 2nd test left something in memory broken).

**Comment by nyh** [comment]
I committed a few patches to fix most of the failing tests. But unfortunately I was able to reproduce sporadically the "preemptable" bug you saw, in just one run out of many. I saw it in tst-without-namespace.so:
```
  TEST tst-without-namespace.so           OSv v0.24-157-gab34727
eth0: 192.168.122.15
Assertion failed: sched::preemptable() (arch/x64/mmu.cc: page_fault: 33)

[backtrace]
0x0000000000225788 <__assert_fail+24>
0x0000000000388cb6 <page_fault+294>
0x0000000000387b36 <???+3701558>
0x00000000003dfdf1 <sched::cpu::do_idle()+161>
0x00000000003e2137 <sched::cpu::idle()+39>
0x00000000003e215b <???+4071771>
0x00000000003e2296 <thread_main_c+38>
0x0000000000388ab5 <???+3705525>
```
I'll try to reproduce this outside the test script, so I could connect with gdb.

**Comment by nyh** [comment]
I was able to reproduce this bug after running tst-without-namespace.so over 600 times successfully (I'm building and running this on Fedora 23).
The backtrace I got from gdb is (release build):
```
#0  0x000000000038f002 in processor::cli_hlt () at arch/x64/processor.hh:247
#1  arch::halt_no_interrupts () at arch/x64/arch.hh:48
#2  osv::halt () at arch/x64/power.cc:24
#3  0x0000000000225747 in abort (
    fmt=fmt@entry=0x68b078 "Assertion failed: %s (%s: %s: %d)\n")
    at runtime.cc:130
#4  0x0000000000225789 in __assert_fail (
    expr=expr@entry=0x7143a3 "sched::preemptable()", 
    file=file@entry=0x70722a "arch/x64/mmu.cc", line=line@entry=33, 
    func=func@entry=0x707250 <page_fault::__func__> "page_fault")
    at runtime.cc:137
#5  0x0000000000388cb7 in page_fault (ef=0xffff800001821068)
    at arch/x64/mmu.cc:33
#6  <signal handler called>
#7  std::__atomic_base<sched::thread*>::load (__m=std::memory_order_relaxed, 
    this=0x0) at /usr/include/c++/5.3.1/bits/atomic_base.h:713
#8  std::atomic<sched::thread*>::load (__m=std::memory_order_relaxed, this=0x0)
    at /usr/include/c++/5.3.1/atomic:416
#9  lockless_queue<sched::thread, &sched::thread::_wakeup_link>::empty (
    this=0xffff9000013230c0) at include/osv/lockless-queue.hh:80
#10 sched::cpu::handle_incoming_wakeups (this=this@entry=0xffff8000012b0040)
    at core/sched.cc:439
#11 0x00000000003dfdf2 in sched::cpu::do_idle (
    this=this@entry=0xffff8000012b0040) at core/sched.cc:405
#12 0x00000000003e2138 in sched::cpu::idle (this=0xffff8000012b0040)
    at core/sched.cc:423
#13 0x00000000003e215c in sched::cpu::<lambda()>::operator() (
    __closure=<optimized out>) at core/sched.cc:165
#14 std::_Function_handler<void(), sched::cpu::init_idle_thread()::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...)
    at /usr/include/c++/5.3.1/functional:1871
#15 0x00000000003e2297 in sched::thread_main_c (t=0xffff80000181c040)
    at arch/x64/arch-switch.hh:164
#16 0x0000000000388ab6 in thread_main () at arch/x64/entry.S:113
```

In frame 10, 
```
(gdb) fr 10
#10 sched::cpu::handle_incoming_wakeups (this=this@entry=0xffff8000012b0040)
    at core/sched.cc:439
439	            while (!q.empty()) {
(gdb) p q
$1 = (lockless_queue<sched::thread, &sched::thread::_wakeup_link> &) @0xffff9000013230c0: {_head = 0x0, _tail = 0xffffa00002b381d0}
```
I don't have a guess yet what is causing this.

**Comment by nyh** [comment]
In a second run, after trying to add a few more assertions in the lockless_queue<> code, I got a  page fault from a slightly different place:

```
#5  0x0000000000388cb7 in page_fault (ef=0xffff8000017e7068)
    at arch/x64/mmu.cc:33
#6  <signal handler called>
#7  sched::cpu::handle_incoming_wakeups (this=this@entry=0xffff80000127a040)
    at core/sched.cc:446
#8  0x00000000003dfe82 in sched::cpu::do_idle (
    this=this@entry=0xffff80000127a040) at core/sched.cc:405
#9  0x00000000003e21c8 in sched::cpu::idle (this=0xffff80000127a040)
    at core/sched.cc:423
```

```
(gdb) fr 7
#7  sched::cpu::handle_incoming_wakeups (this=this@entry=0xffff80000127a040)
    at core/sched.cc:446
446	                } else if (t.tcpu() != this) {
```

So in this run, q.empty() was fine, but q.front() returned garbage (perhaps a null pointer but the variables involved have been optimized out, I'll need to run again in a debug run to get more information).

It is beginning to look like we might have a race somehow, either a bug in the lockless_queue implementation, or a bug in its usage (e.g., two different threads writing to the same queue)?

**Comment by nyh** [comment]
I thought of one possible race. I don't know if it's the one we're seeing in this test, but I wanted to write it down before I forget :-)

It is possible that a thread went into a "waiting" state, and before actually yielding cancels its wait (because it notices the criterion it wanted to wait for was already true). What if somebody wakes this thread between these steps (prepare_wait() and stop_wait())? Normally this would be fine - at worst it would be a spurious wakeup. But what if the thread that stopped waiting decides to exit? The thread object will be destroyed, causing the wakeup queue this thread is in to be messed up (it uses the thread's "_wakeup_link" for linking) and also code in handle_incoming_wakeups() which tries to use "t" can crash (as in the trace above in the call to t.tcpu()).

Looking in the above crash in gdb with "osv info threads", I can't confirm the above suspicion (which might be a separate bug we didn't yet see...). I see the following application threads:

```
 246 (0xffff8000034f9050) tests/tst-witho cpu0 status::waiting sched::thread::join() at core/sched.cc:1274 vruntime  1.10121e-17
	joining on 0xffff8000035f7050
 269 (0xffff8000035f7050) >tests/tst-with cpu0 status::waiting sched::thread::join() at core/sched.cc:1274 vruntime  1.34517e-17
	joining on 0x2000002faad0
 272 (0x2000002faad0) >>tests/tst-wit cpu1 status::waking ?? at arch/x64/entry.S:112 vruntime            0
```

Here thread 272 was just created, with its IP set to the thread_main (line 112 of entry.S; see thread::init_stack() which sets this _state.rip), and then thread::start() was called, which called thread::wake() - which led to use seeing the "status::waking" - and if there is no other waking thread (I'll need to be run the debugging version to be sure) might have led to the crash. Now to only figure out why...

**Comment by nyh** [comment]
I was able to reproduce this bug with a debug build. Got the same as the second crash (at t.tcpu()).
The thread whose wakeup is crashing is indeed the newly started thread (not an exiting thread).

The page fault in t.tcpu() has cr2 = 8, suggesting that when t.tcpu() was called and used "_detached_state->_cpu", _detached_state was still a null pointer. But in gdb, I already see _detached_state having a correct, non-null value. Could this be a memory ordering issue?

**Comment by nyh** [comment]
I think this is unlikely to be a memory ordering issue because:

1. _detached_state is set in thread::thread(), significantly earlier than the thread::start() call which causes the failing wake.
2. By default, we start new threads on the same CPU as the one that started them - they only move later by the load balancer (when they are already running).

Howver, considering the memory barriers in the code:

1. wake() sets status to waking with CST memory ordering but handle_incoming_wakeups() does not read the status, so maybe this doesn't ensure memory ordering.
2. More importantly, wake() uses "release" memory ordering to write to the queue, and the queue read code uses "consume" memory ordering to read from it. But maybe we should have used "acquire", not "consume"? "consume" should have been enough (we get the thread pointer from the queue and then follow pointers inside it) but I'm not sure it is actually implemented correctly in gcc. That being said, I'm not even sure that acquire/release are not no-ops in this case anyway.

**Comment by nyh** [comment]
Sorry for the spam ;-) Another gdb detail which might be relevant: The middle application thread - the one creating the thread whose start is crashing and also waiting for it to end - has a stack trace which looks like this (with "osv thread ..." and "where"):

```
#0  sched::thread::switch_to (this=0xffff800002136040)
    at arch/x64/arch-switch.hh:75
#1  0x00000000005ba89a in sched::cpu::reschedule_from_interrupt (
    this=0xffff800001b36040, called_from_yield=false, preempt_after=...)
    at core/sched.cc:339
#2  0x00000000005ba15c in sched::cpu::schedule () at core/sched.cc:228
#3  0x00000000005bdca7 in sched::thread::wait (this=0xffff8000041ae050)
    at core/sched.cc:1177
#4  0x00000000005c298f in sched::thread::do_wait_until<sched::noninterruptible, sched::thread::dummy_lock, sched::thread::join()::<lambda()> >(sched::thread::dummy_lock &, sched::thread::<lambda()>) (mtx=..., pred=...)
    at include/osv/sched.hh:1033
#5  0x00000000005c1365 in sched::thread::wait_until<sched::thread::join()::<lambda()> >(sched::thread::<lambda()>) (pred=...) at include/osv/sched.hh:1044
#6  0x00000000005be01e in sched::thread::join (this=0x2000002fa930)
    at core/sched.cc:1274
#7  0x000000000056fd60 in osv::rcu_flush () at core/rcu.cc:248
#8  0x0000000000402aaa in elf::program::remove_object (this=0xffff900001fd5000, 
    ef=0xffffa000033a0c00) at core/elf.cc:1194
#9  0x0000000000401c68 in elf::program::<lambda(elf::object*)>::operator()(elf::object *) const (__closure=0xffffa000033f8bf0, obj=0xffffa000033a0c00)
    at core/elf.cc:1136
#10 0x0000000000405fc8 in std::_Sp_counted_deleter<elf::file*, elf::program::load_object(std::__cxx11::string, std::vector<std::__cxx11::basic_string<char> >, std::vector<std::shared_ptr<elf::object> >&)::<lambda(elf::object*)>, std::allocator<void>, (__gnu_cxx::_Lock_policy)2u>::_M_dispose(void) (this=0xffffa000033f8be0)
    at /usr/include/c++/5.3.1/bits/shared_ptr_base.h:466
#11 0x000000000020e2f2 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0xffffa000033f8be0)
    at /usr/include/c++/5.3.1/bits/shared_ptr_base.h:150
#12 0x000000000020bdf1 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x2000002ffda8, __in_chrg=<optimized out>)
    at /usr/include/c++/5.3.1/bits/shared_ptr_base.h:659
#13 0x000000000040646c in std::__shared_ptr<elf::object, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x2000002ffda0, __in_chrg=<optimized out>)
    at /usr/include/c++/5.3.1/bits/shared_ptr_base.h:925
#14 0x000000000063dc7d in std::__shared_ptr<elf::object, (__gnu_cxx::_Lock_policy)2>::reset (this=0xffffa00003355e98)
    at /usr/include/c++/5.3.1/bits/shared_ptr_base.h:1022
#15 0x0000000000638bf1 in osv::application::join (this=0xffffa00003355e10)
    at core/app.cc:222
#16 0x00000000006234e9 in osv::run (path="/tests/payload-env.so", 
    args=std::vector of length 1, capacity 1 = {...}, 
    return_code=0x2000002ffe5c, new_program=false, env=0x0) at core/run.cc:12
#17 0x0000100000c0365b in run_environment_payload (new_program=<optimized out>)
    at /home/nyh/osv.tmp2/osv/tests/env-common.inc:23
#18 0x000000000076ad80 in execute_native_thread_routine ()
#19 0x0000000000699e77 in pthread_private::pthread::<lambda()>::operator()(void) const (__closure=0xffffa00003355a00) at libc/pthread.cc:101
#20 0x000000000069c48a in std::_Function_handler<void(), pthread_private::pthread::pthread(void* (*)(void*), void*, sigset_t, const pthread_private::thread_attr*)::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...)
    at /usr/include/c++/5.3.1/functional:1871
#21 0x000000000044c4d0 in std::function<void ()>::operator()() const (
    this=0xffff8000041ae080) at /usr/include/c++/5.3.1/functional:2267
#22 0x00000000005bdc88 in sched::thread::main (this=0xffff8000041ae050)
    at core/sched.cc:1171
#23 0x00000000005b9cbe in sched::thread_main_c (t=0xffff8000041ae050)
    at arch/x64/arch-switch.hh:164
#24 0x0000000000489036 in thread_main () at arch/x64/entry.S:113
```

So osv::run() called already called osv::application::join() which apparently decided to delete the shared object because it isn't in use any more. I'm not sure this was done in the right time (I'll open a separate issue about that), but the more interesting thing we can learn from this backtrace is that  in frame 6 we discover the actual identity of the thread whose startup is crashing: It's not the regular application thread, it's a temporary thread created by osv::rcu_flush().

Another interesting discovery here is that the osv::rcu_flush() code can wake a thread on a different core. Looking in GDB, we discover that indeed the problem happens when a thread on sched::cpus[1] starts a thread on c=sched::cpus[1]. So this brings back the possibility of memory ordering issues.

**Comment by nyh** [comment]
Changing memory order "consume" to "acquire" in lockless queue did not fix the problem, but seems to have changed it. Now I see a failure on the following assert:
```c++
template <class T, lockless_queue_link<T> T::*link>
void lockless_queue<T, link>::pop_front()
{
    auto elem = _head->_next.load(std::memory_order_acquire);
    _head = (elem->*link)._next;
    assert(_head);
}
```

(having a null _head would be bad because it will be dereferences next time we check empty()).

However, when connecting with GDB at the point of the assert(), I see that:
```
(gdb) p elem->_wakeup_link 
$6 = {_helper = 0xffffa00002769a58, _next = 0xffffa00002769af8}
```
So clearly _head, set to, elem->_wakeup_link->_next, should not actually be zero! Again looks like we have a memory ordering issue or some sort of race (like multiple writers using the same queue?).

**Comment by nyh** [comment]
BTW, I reverted commit bca7bbbe which re-implemented the lockless-queue.hh wakeup queue, and this test still fails (after several hundred successes). So it's probably not a bug in lockless-queue.hh. And it's also not an issue with multiple writers (which the previous implementation actually allowed, even if unnecessarily).

**Comment by nyh** [comment]
I'm closing this issue and opening a new one for just for the remaining failure of tst-without-namespace.so with the relevant information.
If you see failures of other tests can you please open separate issues?
